src/cpp/astar.cpp

```diff
@@ -27,15 +27,33 @@ bool operator<(const Node &n1, const Node &n2) {
 // See for various grid heuristics:
 // http://theory.stanford.edu/~amitp/GameProgramming/Heuristics.html#S7
 // L_\inf norm (diagonal distance)
-inline float linf_norm(int i0, int j0, int i1, int j1) {
-  return std::max(std::abs(i0 - i1), std::abs(j0 - j1));
+inline float linf_norm(int current_x, int current_y, int goal_x, int goal_y) {
+  return std::max(std::abs(current_x - goal_x), std::abs(current_y - goal_y));
 }
 
 // L_1 norm (manhattan distance)
-inline float l1_norm(int i0, int j0, int i1, int j1) {
-  return std::abs(i0 - i1) + std::abs(j0 - j1);
+inline float l1_norm(int current_x, int current_y, int goal_x, int goal_y) {
+  return std::abs(current_x - goal_x) + std::abs(current_y - goal_y);
 }
 
+// Orthogonal x (moves by x first)
+inline float l1_orthogonal_x(int current_x, int current_y, int goal_x, int goal_y) {
+  return std::abs(current_x - goal_x);
+}
+
+// Orthogonal y (moves by y first)
+inline float l1_orthogonal_y(int current_x, int current_y, int goal_x, int goal_y) {
+  return std::abs(current_y - goal_y);
+}
+
+// tie breaker (prefer straight paths to goal)
+inline float tie_breaker_func(int current_x, int current_y, int goal_x, int goal_y, int start_x, int start_y) {
+  int dx1 = current_x - goal_x;
+  int dy1 = current_y - goal_y;
+  int dx2 = start_x - goal_x;
+  int dy2 = start_y - goal_y;
+  return std::abs(dx1*dy2 - dx2*dy1);
```

Comment:
This value could be arbitrarily large based on the size of the grid, right? Does this break the admissibility of the heuristic?

Reply:
According to http://theory.stanford.edu/~amitp/GameProgramming/Heuristics.html, the tie breaker logic is still optimal.

Not sure how one could prove otherwise, the very small coefficient number suggests it may not be optimal for all coefficient numbers.